File: android/src/main/java/xiuyue/tech/plugins/clipboard/YueClipboardPlugin.java
package xiuyue.tech.plugins.clipboard;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.List;

@CapacitorPlugin(name = "YueClipboard")
public class YueClipboardPlugin extends Plugin {

    // Initialize implementation with the context from the plugin
    private YueClipboard implementation;

    @Override
    public void load() {
        super.load();
        // Initialize the implementation when the plugin loads
        implementation = new YueClipboard(getContext());
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");
        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void write(PluginCall call) {
        String text = call.getString("string");
        String image = call.getString("image");
        String url = call.getString("url");

        try {
            implementation.write(text, image, url);
            call.resolve();
        } catch (YueClipboard.ClipboardException e) {
            call.reject("CapacitorException: " + e.getMessage());
        } catch (Exception e) {
            call.reject("CapacitorException: Failed to write to clipboard: " + e.getMessage());
        }
    }

    @PluginMethod
    public void read(PluginCall call) {
        try {
            YueClipboard.ClipboardReadResult result = implementation.read();
            JSObject ret = new JSObject();
            ret.put("value", result.value);
            ret.put("type", result.type);
            call.resolve(ret);
        } catch (YueClipboard.ClipboardException e) {
            call.reject("CapacitorException: " + e.getMessage());
        } catch (Exception e) {
            call.reject("CapacitorException: Failed to read from clipboard: " + e.getMessage());
        }
    }

    @PluginMethod
    public void hasContent(PluginCall call) {
        try {
            boolean hasContent = implementation.hasContent();
            JSObject ret = new JSObject();
            ret.put("hasContent", hasContent);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("CapacitorException: Failed to check clipboard content: " + e.getMessage());
        }
    }

    @PluginMethod
    public void clear(PluginCall call) {
        try {
            implementation.clear();
            call.resolve();
        } catch (YueClipboard.ClipboardException e) {
            call.reject("CapacitorException: " + e.getMessage());
        } catch (Exception e) {
            call.reject("CapacitorException: Failed to clear clipboard: " + e.getMessage());
        }
    }

    @PluginMethod
    public void getAvailableTypes(PluginCall call) {
        try {
            List<String> types = implementation.getAvailableTypes();
            JSArray typesArray = new JSArray();
            for (String type : types) {
                typesArray.put(type);
            }
            
            JSObject ret = new JSObject();
            ret.put("types", typesArray);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("CapacitorException: Failed to get available types: " + e.getMessage());
        }
    }
}

File: ios/Sources/YueClipboardPlugin/YueClipboardPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(YueClipboardPlugin)
public class YueClipboardPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "YueClipboardPlugin"
    public let jsName = "YueClipboard"

    // Define all methods that will be exposed to JavaScript
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "write", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "read", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasContent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAvailableTypes", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = YueClipboard()

    // MARK: - Plugin Method Implementations

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    @objc func write(_ call: CAPPluginCall) {
        let text = call.getString("string")
        let image = call.getString("image")
        let url = call.getString("url")

        do {
            try implementation.write(text: text, image: image, url: url)
            call.resolve()
        } catch let error as YueClipboard.ClipboardError {
            call.reject(error.localizedDescription)
        } catch {
            call.reject("Failed to write to clipboard: \(error.localizedDescription)")
        }
    }

    @objc func read(_ call: CAPPluginCall) {
        do {
            let result = try implementation.read()
            call.resolve([
                "value": result.value,
                "type": result.type
            ])
        } catch let error as YueClipboard.ClipboardError {
            call.reject(error.localizedDescription)
        } catch {
            call.reject("Failed to read from clipboard: \(error.localizedDescription)")
        }
    }

    @objc func hasContent(_ call: CAPPluginCall) {
        let hasContent = implementation.hasContent()
        call.resolve(["hasContent": hasContent])
    }

    @objc func clear(_ call: CAPPluginCall) {
        do {
            try implementation.clear()
            call.resolve()
        } catch let error as YueClipboard.ClipboardError {
            call.reject(error.localizedDescription)
        } catch {
            call.reject("Failed to clear clipboard: \(error.localizedDescription)")
        }
    }

    @objc func getAvailableTypes(_ call: CAPPluginCall) {
        let types = implementation.getAvailableTypes()
        call.resolve(["types": types])
    }
}

